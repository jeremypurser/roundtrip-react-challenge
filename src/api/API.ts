export interface Alias {
  plan_name: string;
  carrier_name: string;
  unmatched_plan_id: number;
  master_plan_id: number;
}

export interface API {
  getUnmatchedPlan: (id: number) => Promise<any>;

  getMasterPlans: () => Promise<any>;

  addAlias: (alias: Alias) => Promise<any>;

  addMasterPlan: (masterPlan: { name: string }) => Promise<any>;
}
