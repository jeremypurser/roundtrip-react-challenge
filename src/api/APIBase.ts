export type HTTPMethod = 'POST' | 'GET' | 'PUT' | 'PATCH' | 'DELETE';

export class APIBase {
  private baseUrl: string;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  private log(message: string) {
    if (process.env.NODE_ENV !== 'production') {
      console.log(message);
    }
  }

  protected _fetch(req: {
    endpoint: string;
    method: HTTPMethod;
    data?: any;
  }): Promise<any> {
    const requestInit: RequestInit = {
      method: req.method,
    };

    if (
      req.method === 'POST' ||
      req.method === 'PUT' ||
      req.method === 'PATCH'
    ) {
      requestInit.headers = { 'Content-Type': 'application/json' };
      requestInit.body = JSON.stringify(req.data);
    }

    this.log(`
    ---- REQUEST ----
    url: ${this.baseUrl}/${req.endpoint}
    method: ${req.method}
    data: ${req.data ? JSON.stringify(req.data, null, 2) : 'N/A'}
    `);

    return fetch(`${this.baseUrl}/${req.endpoint}`, requestInit)
      .then(response => response.json())
      .then(json => {
        this.log(`
        ---- RESPONSE ----
        response: ${JSON.stringify(json, null, 2)}
        `);

        return json;
      });
  }
}
