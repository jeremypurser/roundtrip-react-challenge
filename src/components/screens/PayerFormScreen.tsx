import * as React from 'react';

export const PayerFormScreen = () => {
  return (
    <form id="payer-form">
      {/* Carrier and plan name */}
      <label htmlFor="fname">First name:</label>
      <input type="text" id="fname" name="fname" />

      {/* Company Id */}
      <label htmlFor="company-id">Company Id:</label>
      <input type="number" id="company-id" name="company-id" />

      {/* Phone number */}
      <label htmlFor="phone">Phone number:</label>
      <input type="tel" id="phone" name="phone" />

      {/* Match */}
      <select name="match" id="match" form="payer-form">
        <option value="example">Example</option>
      </select>

      <button type="button">Match</button>
      <button type="button">Create Insurance</button>
    </form>
  );
};
