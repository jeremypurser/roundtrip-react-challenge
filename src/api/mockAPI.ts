import { Alias, API } from './API';

export const mockAPIClient: API = {
  getUnmatchedPlan: (id: number) =>
    new Promise(resolve => {
      resolve({
        id,
        plan_name: 'Gold Plan',
        carrier_name: '1199 National Benefit Fund',
        unmatched_plan_id: 1,
        master_plan_id: 821,
      });
    }),

  getMasterPlans: () =>
    new Promise(resolve => {
      resolve([
        {
          id: 818,
          name: '1199SEIU Home Care Benefit Fund',
        },
        {
          id: 819,
          name: '1199SEIU Home Health Aide Benefit Fund',
        },
        {
          id: 8195,
          name: '1199SEIU Licensed Practical Nurses Welfare Fund',
        },
        {
          id: 821,
          name: '1199SEIU National Benefit Fund',
        },
      ]);
    }),

  addAlias: (alias: Alias) =>
    new Promise(resolve => {
      resolve({
        alias,
      });
    }),

  addMasterPlan: (masterPlan: { name: string }) =>
    new Promise(resolve => {
      resolve(masterPlan);
    }),
};
