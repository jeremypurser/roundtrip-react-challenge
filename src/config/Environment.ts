import { API } from '../api/API';

interface PaymentEnvironment {
  api: API;
}

export class Environment {
  api: API;

  static current: PaymentEnvironment;

  static set(init: PaymentEnvironment) {
    if (this.current === undefined) {
      this.current = new this(init);
    }
  }

  private constructor(init: PaymentEnvironment) {
    this.api = init.api;
  }
}
