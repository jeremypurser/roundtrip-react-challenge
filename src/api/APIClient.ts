import { Alias, API } from './API';
import { APIBase } from './APIBase';

export class APIClient extends APIBase implements API {
  /**
   * endpoint -> /unmatched-plans/:id
   */
  getUnmatchedPlans(id: number): Promise<any> {
    return this._fetch({
      endpoint: `unmatched-plans/${id}`,
      method: 'GET',
    });
  }

  /**
   * endpoint -> master-plans
   */
  getMasterPlans(): Promise<any> {
    return this._fetch({
      endpoint: 'master-plans',
      method: 'GET',
    });
  }

  /**
   * endpoint -> /aliases
   * payload -> {
      "id":1,
      "plan_name":"Faker Plan",
      "carrier_name":"Roundtrip Insurance Company",
      "unmatched_plan_id":2,
      "master_plan_id":1
    }
   */
  addAlias(alias: Alias): Promise<any> {
    return this._fetch({
      endpoint: 'aliases',
      method: 'POST',
      data: alias,
    });
  }
}
