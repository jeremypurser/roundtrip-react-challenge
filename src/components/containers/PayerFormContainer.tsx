import * as React from 'react';
import Spinner from 'react-bootstrap/Spinner';
import { Environment } from '../../config/Environment';

const { useEffect, useState } = React;

export interface PayerFormProps {
  plans: {
    id: number;
    name: string;
  }[];
}

/**
 * Higher Order Component that connects purely presentational component to state/business logic
 * @param Screen PayerFormScreen
 * @returns Connected Component
 */
export const payerFormContainer = (
  Screen: React.ComponentType<PayerFormProps>
) => () => {
  const { api } = Environment.current;

  const [loading, setLoading] = useState(false);

  const [masterPlans, setMasterPlans] = useState<
    { id: number; name: string }[]
  >([]);

  useEffect(() => {
    // make API calls
    setLoading(true);
    api
      .getMasterPlans()
      .then(setMasterPlans)
      .catch(error => {
        // ALERT
      })
      .finally(() => setLoading(false));
  }, [api]);

  return loading ? (
    <Spinner animation="border" role="status">
      <span className="sr-only">Loading...</span>
    </Spinner>
  ) : (
    <Screen plans={masterPlans} />
  );
};
