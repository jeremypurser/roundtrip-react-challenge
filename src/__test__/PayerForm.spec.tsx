import React from 'react';
import renderer from 'react-test-renderer';
import { mockAPIClient } from '../api/mockAPI';
import { payerFormContainer } from '../components/containers/PayerFormContainer';
import { PayerFormScreen } from '../components/screens/PayerFormScreen';
import { Environment } from '../config/Environment';

Environment.set({
  api: mockAPIClient,
});

const PayerForm = payerFormContainer(PayerFormScreen);

describe('PayerForm', () => {
  const { api } = Environment.current;

  it('should render correctly', () => {
    const tree = renderer.create(<PayerForm />).toJSON();

    Promise.all([api.getMasterPlans(), api.getUnmatchedPlan(1)]).then(() => {
      expect(tree).toMatchSnapshot();
    });
  });
});
