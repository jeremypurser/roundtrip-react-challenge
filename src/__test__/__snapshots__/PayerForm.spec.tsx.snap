// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayerForm should render correctly 1`] = `
Array [
  <h2>
     
  </h2>,
  <form
    className=""
  >
    <div
      className="mb-3 input-group"
    >
      <div
        className="input-group-prepend"
      >
        <span
          className="input-group-text"
          id="carrier-and-plan-name"
        >
          Carrier and plan name:
        </span>
      </div>
      <input
        aria-describedby="carrier-and-plan-name"
        aria-label="Carrier and plan name"
        className="form-control"
        readOnly={true}
        type="text"
        value="undefined undefined"
      />
    </div>
    <div
      className="mb-3 input-group"
    >
      <div
        className="input-group-prepend"
      >
        <span
          className="input-group-text"
          id="company-id"
        >
          Company Id:
        </span>
      </div>
      <input
        aria-describedby="company-id"
        aria-label="Company Id"
        className="form-control"
        readOnly={true}
        type="text"
      />
    </div>
    <div
      className="mb-3 input-group"
    >
      <div
        className="input-group-prepend"
      >
        <span
          className="input-group-text"
          id="company-phone"
        >
          Phone number:
        </span>
      </div>
      <input
        aria-describedby="company-phone"
        aria-label="Phone number"
        className="form-control"
        readOnly={true}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <select
        className="form-control"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Select
        </option>
      </select>
    </div>
    <button
      className="btn btn-light"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginRight": "1rem",
        }
      }
      type="button"
    >
      Match
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Insurance
    </button>
  </form>,
]
`;
